File: Google/846. Hand of Straights.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetCode
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var solution = new Solution();
            solution.IsNStraightHand(new int[] { 1, 2, 3, 5, 5, 6, 7 }, 3);
        }
    }

    public class Solution
    {
        public bool IsNStraightHand(int[] hand, int groupSize)
        {
            if (hand.Length % groupSize != 0) return false;

            Array.Sort(hand);

            var counter = new Dictionary<int, int>();
            foreach (var num in hand)
            {
                if (counter.ContainsKey(num))
                    counter[num]++;
                else
                    counter[num] = 1;
            }

            foreach (var num in hand)
            {
                if (counter[num] == 0) continue;

                counter[num]--;
                for (var i = 1; i < groupSize; i++)
                {
                    var value = num + i;
                    if (counter.ContainsKey(value) && counter[value] > 0)
                        counter[value]--;
                    else
                        return false;
                }
            }

            return true;
        }
    }
}

File: Google/846. Hand of Straights.js
/**
 * @param {number[]} hand
 * @param {number} groupSize
 * @return {boolean}
 */
var isNStraightHand = function(hand, groupSize) {
	if (hand.length % groupSize !== 0) {
    	return false;
    }

    let counter = new Map();
    for (let num of hand) {
    	if (counter.has(num)) {
    		counter.set(num, counter.get(num) + 1);
    	} else {
    		counter.set(num, 1);
    	}
    }

    let sorted = [...counter.keys()].sort((a, b) => a - b);
    for (let num of sorted) {
    	let cnt = counter.get(num);

    	if (cnt === 0) continue;

    	for (let i = 1; i < groupSize; i++) {
    		let next = num + i;

    		if (!counter.has(next) || counter.get(next) < cnt) {
    			return false;
    		}

    		counter.set(next, counter.get(next) - cnt);
    	}
    }

    return true;
};

isNStraightHand([1,2,3,6,2,3,4,7,8], 3);